### PR TITLE
Update JS bindings with CrossSection + reorganize

### LINKED ISSUE #360: Add JS bindings for CrossSection
There's also a few other APIs missing, like `Split` and `Trim`.

### PR DESCRIPTION
**Pull request by geoffder**
Fixes #360

- add CrossSection class
- move static method (constructors) under their respective classes
- add Manifold.{split, splitByPlane}
- enable usage of `instanceof` on `Manifold` (and now `CrossSection`) objects
- simplify registry indirection in `worker.ts`, enabling method de-structuring (opening static function names out of class module name-spaces)

Build changes:
- break C++ helpers out from `bindings.cpp` to `helpers.cpp`
- ensure that `ts` files are copied during builds even when nothing is recompiled

Note that in addition to the new bindings, the most of the existing toplevel functions have been moved under their respective classes as static methods (breaking change!). I haven't added any new examples yet, but the existing ones are passing and do make use of `CrossSection` now (some transparently, and some directly), but I think enough of this is together that review is worthwhile.

### CODE REVIEW

**Comment by geoffder** [comment]
Ah looks like the previously [mentioned](https://github.com/elalish/manifold/issues/360#issuecomment-1563435673) addition to `debug` in `worker.ts` must have been dropped while I was resolving conflicts, woops. For reference, I think it might be helpful to allow `show` and `only` to work on `CrossSection` as well, similar to how 2d shapes show up in OpenSCAD's preview mode.

```javascript
function debug(shape: Manifold | CrossSection, map: Map<number, Mesh>) {
  const manifold = (shape instanceof module.CrossSection) 
                   ? shape.extrude(1).translate([0, 0, -0.5]) 
                   : shape;
  const result = manifold.asOriginal();
  map.set(result.originalID(), result.getMesh());
  return result;
};
```

Also, after looking at it again just now, if the cleanup step is required to make sure that Manifolds get freed, is `debug` currently (even without this addition) leaking the result of `asOriginal()` (in need of a   `memoryRegistry.push(result)`)?

**Comment by elalish** [comment]
Good question - I *think* that gets cleaned up by the modified `asOriginal` function like the rest, but let me know if I'm missing something. I like the cross-section debug idea - how about extrude 1/100 of the max dimension to make it scale invariant?

File: bindings/wasm/examples/worker.ts
Comment (elalish) [inline bindings/wasm/examples/worker.ts:15]: Can we figure out what's going on with our clang formats? It's impossible to read these diffs if they're full of white space corrections. Are you using VSCode? I have a `.clang-format` file in our root, so ideally that should be standardizing this. I should really figure out how to add a format check to our CI...

File: bindings/wasm/examples/worker.ts
Comment (elalish) [inline bindings/wasm/examples/worker.ts:33]: Were you getting an error here?

File: bindings/wasm/manifold.d.ts
Comment (elalish) [inline bindings/wasm/manifold.d.ts:47]: Intensional?

File: bindings/wasm/helpers.cpp
Comment (elalish) [inline bindings/wasm/helpers.cpp:13]: Good idea splitting this file out. 👍

**Review by elalish — CHANGES_REQUESTED** [review]
Looking pretty good, but let's clean up the format so I can do a better pass. The examples especially will break pretty badly, since their white space gets processed.

File: bindings/wasm/examples/worker.ts
Comment (geoffder) [reply bindings/wasm/examples/worker.ts:33]: The CI errored out yea (see the first commit of the PR)

File: bindings/wasm/examples/worker.ts
Comment (geoffder) [reply bindings/wasm/examples/worker.ts:15]: Yea sorry about that, I'm using the `ts-lsp` in emacs. The clang-lsp has always lined up with your formatting setup without any tweaking for me, but the js formatting has other ideas on my end.

File: bindings/wasm/manifold.d.ts
Comment (geoffder) [reply bindings/wasm/manifold.d.ts:47]: That'd be autoformatting

Comment (elalish) [reply bindings/wasm/manifold.d.ts:47]: No, you lost a `?` I added in my last PR.

**Comment by geoffder** [comment]
> Good question - I _think_ that gets cleaned up by the modified `asOriginal` function like the rest, but let me know if I'm missing something. 

This doesn't go through the modified version does it? It comes before the indirection is set up.maybe if it was moved down the file.

> I like the cross-section debug idea - how about extrude 1/100 of the max dimension to make it scale invariant?

Sure, that doesn't sound bad.

File: bindings/wasm/manifold.d.ts
Comment (geoffder) [reply bindings/wasm/manifold.d.ts:47]: Oh then it's my fault when I was resolving merge conflicts.

File: bindings/wasm/examples/worker.ts
Comment (elalish) [reply bindings/wasm/examples/worker.ts:33]: No, that's from the missing `?` below.

Comment (geoffder) [reply bindings/wasm/examples/worker.ts:33]: Hmm well the CI passed once that comment was added

**Comment by elalish** [comment]
> This doesn't go through the modified version does it? It comes before the indirection is set up.maybe if it was moved down the file.

Oh, good catch! Let's do that.

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [inline bindings/wasm/examples/public/examples.js:507]: Intentional? Have you tested ManifoldCAD.org locally?

File: bindings/wasm/bindings.cpp
Comment (elalish) [inline bindings/wasm/bindings.cpp:171]: intentional?

File: bindings/wasm/bindings.cpp
Comment (elalish) [inline bindings/wasm/bindings.cpp:173]: 👍

File: bindings/wasm/examples/vite.config.js
Comment (geoffder) [inline bindings/wasm/examples/vite.config.js:1]: Ah missed a re-reformat.

File: bindings/wasm/bindings.js
Comment (elalish) [inline bindings/wasm/bindings.js:111]: Doesn't `vararg2vec2` already take care of this case?

File: bindings/wasm/examples/public/examples.js
Comment (geoffder) [inline bindings/wasm/examples/public/examples.js:506]: Was this something you added in your recent commit that I missed again? Will add back.

File: bindings/wasm/examples/worker.ts
Comment (geoffder) [inline bindings/wasm/examples/worker.ts:135]: Woops, flagged for deletion

File: bindings/wasm/bindings.js
Comment (elalish) [inline bindings/wasm/bindings.js:491]: Why `ofMesh` instead of just the plain constructor?

File: bindings/wasm/bindings.js
Comment (elalish) [inline bindings/wasm/bindings.js:495]: Necessary?

File: bindings/wasm/bindings.js
Comment (elalish) [inline bindings/wasm/bindings.js:589]: Should this be on the `CrossSection` prototype instead?

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [inline bindings/wasm/examples/public/examples.js:121]: Maybe we should destructure at the top: `const {union, cylinder} = Manifold;` so we don't have to add so much verbosity?

**Review by elalish — CHANGES_REQUESTED** [review]
Looking great, thanks! Would you mind pulling master so we can double-check the formatting?

File: bindings/wasm/helpers.cpp
Comment (geoffder) [reply bindings/wasm/helpers.cpp:201]: As far as my searches could tell, emscripten does not take care of std::pair automatically, so this just transforms into a vector which can easily be mapped into a js array (as with decompose).

File: bindings/wasm/bindings.js
Comment (geoffder) [reply bindings/wasm/bindings.js:589]: The triangulation is mostly used for building meshes though right? Triangulating a polygon doesn't do much good if your goal is using it as a CrossSection.

File: bindings/wasm/bindings.js
Comment (geoffder) [reply bindings/wasm/bindings.js:491]: I feel like `Manifold.ofMesh(m)` is clearer than `new Manifold(m)` (same with the CrossSection constructor alias I added, `ofPolygons`). Also, it stands out as an available option if it is there in the list of functions imo (with a hint of what you need to give it in the name). The regular constructor still exists too.

File: bindings/wasm/examples/public/examples.js
Comment (geoffder) [reply bindings/wasm/examples/public/examples.js:507]: No I haven't yet since I need to install a browser other than Firefox. It wasn't intentional though, as mentioned in the review I added to this block.

File: bindings/wasm/bindings.js
Comment (elalish) [reply bindings/wasm/bindings.js:589]: I just mean the input is a CrossSection, not the output. I was imagining something like `const triangles = polygon.triangulate();` instead of `const triangles = Module.triangulate(polygon);`.

File: bindings/wasm/bindings.js
Comment (elalish) [reply bindings/wasm/bindings.js:491]: Alright, that sounds reasonable.

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [reply bindings/wasm/examples/public/examples.js:507]: Firefox should work too, with a small tweak: https://github.com/elalish/manifold/issues/328#issuecomment-1473847102

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [reply bindings/wasm/examples/public/examples.js:506]: yeah, thanks.

File: bindings/wasm/bindings.js
Comment (geoffder) [reply bindings/wasm/bindings.js:589]: Ah right. The issue I can think of is that the points Cana change from the users original polygon on making a cross section, so they'll have to make it, put it through triangulate, then get the points from cross section so that they match up with what triangulate worked on. Depends on how low level we want to keep the API for triangulate (and its use case is generally pretty low level, making bespoke meshes).

Comment (elalish) [reply bindings/wasm/bindings.js:589]: Okay, good point. Let's come back to this another time then.

File: bindings/wasm/examples/public/examples.js
Comment (geoffder) [reply bindings/wasm/examples/public/examples.js:507]: That did work, thanks! Maybe we should mention that in a README somewhere.

File: bindings/wasm/bindings.cpp
Comment (geoffder) [reply bindings/wasm/bindings.cpp:173]: Yes, the `class_function` method of embind was not working for me for setting up the static methods (compiles, but then the function is not actually available in `examples.js` (function undefined). Maybe because of the way that the object is being replaced in `binding.js`?

File: bindings/wasm/bindings.js
Comment (geoffder) [reply bindings/wasm/bindings.js:497]: Yes, the class_function method of embind was not working for me for setting up the static methods (compiles, but then the function is not actually available in examples.js (function undefined). Maybe because of the way that the object is being replaced in binding.js? I imagine there is some reorganization that would not require this, but the way we have it setup right now it doesn't work.

File: bindings/wasm/bindings.cpp
Comment (geoffder) [reply bindings/wasm/bindings.cpp:173]: See https://github.com/elalish/manifold/pull/440#discussion_r1207452143

File: bindings/wasm/examples/public/examples.js
Comment (geoffder) [reply bindings/wasm/examples/public/examples.js:121]: Within the example body (where they will appear to the user when viewing the example), or above? If it was at the top, that pollute the whole namespace for the editor I guess. The shape constructors have no collision between Manifold and CrossSection if there were opened out, but the booleans would. Not sure what the best course is.

File: bindings/wasm/bindings.js
Comment (geoffder) [reply bindings/wasm/bindings.js:111]: Ah perhaps, I just copied what `Manifold.prototype.scale` is doing.

**Comment by geoffder** [comment]
Gyroid example just refusing to not timeout it seems.

File: bindings/wasm/bindings.js
Comment (elalish) [reply bindings/wasm/bindings.js:498]: Interesting, thanks.

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [reply bindings/wasm/examples/public/examples.js:121]: In the body where they appear to the user. Basically the same pattern I use with `const {vec3} = glMatrix;`.

File: bindings/wasm/examples/vite.config.js
Comment (elalish) [inline bindings/wasm/examples/vite.config.js:6]: It's timing out because of a rebase issue - this is the wrong key, so we're using the default of 5000.

File: bindings/wasm/bindings.js
Comment (elalish) [reply bindings/wasm/bindings.js:111]: Never a bad time to clean up :)

**Review by elalish — APPROVED** [review]
Looks like this is ready once the tests pass. Thanks! If you click the merge button, please check when deployment completes and then test ManifoldCAD.org. Since we don't have any staging, I'd like to catch problems early.

File: bindings/wasm/examples/public/examples.js
Comment (geoffder) [reply bindings/wasm/examples/public/examples.js:121]: Ok I'll hunt through for places where it looks like it could use a local opening like that.

File: bindings/wasm/examples/vite.config.js
Comment (geoffder) [reply bindings/wasm/examples/vite.config.js:6]: Ahhh ok, I thought I picked the right one on that by obviously I was going a bit crosseyed going through the autoformat infected diffs (my own fault of course).

**Comment by geoffder** [comment]
> Looks like this is ready once the tests pass. Thanks! If you click the merge button, please check when deployment completes and then test ManifoldCAD.org. Since we don't have any staging, I'd like to catch problems early.

Seems like most things are resolved, but I haven't touched the interface files yet which are giving inaccurate/incomplete completion information with all of the static function reorganization and the CrossSection addition unaccounted for.

File: bindings/wasm/CMakeLists.txt
Comment (elalish) [inline bindings/wasm/CMakeLists.txt:23]: Probably, but my CMake foo is weak. Feel free to play around!

**Comment by geoffder** [comment]
Ah looks like I should have tested the scale cleanup before committing, might have been what broke here.

File: bindings/wasm/CMakeLists.txt
Comment (geoffder) [reply bindings/wasm/CMakeLists.txt]: I believe that I've got something working for the interface copying that does not require rebuilding the bindings needlessly. :+1:

**Review by elalish — CHANGES_REQUESTED** [review]

File: bindings/wasm/examples/worker.ts
Comment (geoffder) [reply bindings/wasm/examples/worker.ts:193]: The box returned in JS is just a couple fields though (not a Box with all the methods).

File: bindings/wasm/bindings.js
Comment (geoffder) [reply bindings/wasm/bindings.js:95]: I did it when I thought that keywords existed (rather than just ordering like in cpp) out of python habit (in ocaml there is name punning, so wouldn't be doing the repetition there). I'll go back and drop it.

File: bindings/wasm/examples/worker.ts
Comment (geoffder) [reply bindings/wasm/examples/worker.ts:198]: I've played with it a bit, it isn't as nice as with the manifolds for obvious reasons as you mention, but it does let you quickly red/ghost them and show them alongside manifolds if you stick the manifold made by `show` in a node for example.

File: bindings/wasm/helpers.cpp
Comment (geoffder) [reply bindings/wasm/helpers.cpp:110]: Ah ok, I'll just pass an int to the helper function from js instead.

File: bindings/wasm/manifold-encapsulated-types.d.ts
Comment (geoffder) [reply bindings/wasm/manifold-encapsulated-types.d.ts:643]: We could, I was just organizing them by purpose. I'm not sure what the way that this interface will normally be viewed.

File: bindings/wasm/examples/worker.ts
Comment (elalish) [reply bindings/wasm/examples/worker.ts:193]: Ah, fair. Should probably add those someday...

**Comment by geoffder** [comment]
As an aside, I have come across a funny bug, the following fails in the editor with `RuntimeError: index out of bounds`
```javascript
const box = CrossSection.square([80, 80], true).offset(10, "Round").extrude(100, 0, 0, [1, 1], false);
```
However, changing the final argument (center) to `true` (which simply translates the manifold down) works.

File: bindings/wasm/manifold-encapsulated-types.d.ts
Comment (elalish) [reply bindings/wasm/manifold-encapsulated-types.d.ts:643]: Probably mostly by other maintainers. Just try to make it obvious where someone would put the next method.

**Comment by elalish** [comment]
> As an aside, I have come across a funny bug, the following fails in the editor with `RuntimeError: index out of bounds`
> 
> ```js
> const box = CrossSection.square([80, 80], true).offset(10, "Round").extrude(100, 0, 0, [1, 1], false);
> ```
> 
> However, changing the final argument (center) to `true` (which simply translates the manifold down) works.

Ugh, WASM errors are so obtuse. Can you repro it in a C++ test? I just fixed some like that for Warp, which turned out to be because I wasn't checking inputs strictly enough.

File: bindings/wasm/examples/worker.ts
Comment (elalish) [reply bindings/wasm/examples/worker.ts:198]: I dunno, I think this could get pretty confusing for users. Let's hold off until we've found a better way.

**Comment by geoffder** [comment]
I did try doing `GetMesh` on the non-centred version in cpp, and it exported fine without crashing :grimacing:.

```cpp
TEST(CrossSection, RoundedBox) {
  auto sq = CrossSection::Square({80, 80}, true)
                .Offset(10, CrossSection::JoinType::Round);
  auto uncentred = Manifold::Extrude(sq, 100, 0, 0, {1, 1});
  auto ball = Manifold::Sphere(60, 100);
  auto mesh = (uncentred - ball).GetMesh();
#ifdef MANIFOLD_EXPORT
  if (options.exportModels)
    ExportMesh("cross_section_rounded_box.glb", mesh, {});
#endif
}
```
It *did* however take almost 37s to run though.

**Comment by elalish** [comment]
Weird, it seems way too simple to take that long. How many facets for "round"? Oh well, we should probably move this discussion to a separate issue anyway.

**Comment by geoffder** [comment]
> Weird, it seems way too simple to take that long. How many facets for "round"? Oh well, we should probably move this discussion to a separate issue anyway.

Ok. It is a pretty weird problem given than a mere translation makes the difference. This use of offset is bringing it out, but where the error is and how insidious I don't know.

**Comment by pca006132** [comment]
I finally have a break now. I just have a look at it, there are 281010 edges in the `uncentred` mesh, so most of the time is spent on `RecursiveSwapEdge` and causing the slowness.

**Comment by pca006132** [comment]
Related: http://www.angusj.com/clipper2/Docs/Units/Clipper.Offset/Classes/ClipperOffset/Properties/ArcTolerance.htm
Perhaps we should have this in CrossSection, similar to what we have with Manifold.

**Comment by geoffder** [comment]
From how smooth the roundover looks I'm not too surprised. 

Calculating a default ourselves based on the delta, rather than leaving it up to the Clipper2 default or the user you mean?

**Comment by pca006132** [comment]
Yeah, we can allow the user to set the `minCircularAngle`/`circularSegments` and calculate it ourselves, or ask the user to specify a `minCircularEdgeLength`. We should probably have the `minCircularAngle` set to 10 or something, so the default performance is good enough.

**Comment by geoffder** [comment]
Caught what I hope to be the last bugs while testing out `CrossSection` in the editor earlier, think it is looking pretty good now.

I remembered a question I'd forgotten around the business in [getManifoldDTS](https://github.com/elalish/manifold/blob/507b8448401ddb899be5cc6ecaaf78392b4da9e1/bindings/wasm/examples/editor.js#L231) in `editor.js`. I've copied the corresponding updated block over what was there before, but I'm wondering what the purpose is, since naively, it seems incomplete. It gets `encapsulated`, but then doesn't seem to be aliased as `T` to work with the subsequent `typeof` declarations, and the top-level types aren't exported as they are in `manifold.d.ts` either. Though, again, I've come into working on this with no familiarity to the js/ts ecosystem, so I could be missing something obvious.

If this stuff is actually (still?) necessary, it seems pretty fragile given that it requires manually copying over the interface into the string block (though it isn't something that happens *too* often I suppose).

**Comment by elalish** [comment]
I believe #348 is where the `getManifoldDTS` came from - linked issues there may give a little background. It's a total hack to work around Monaco not liking `.d.ts` files that import others. I don't have a chance to review right now - it's entirely possible it is now out of date or there's a better way.

**Comment by geoffder** [comment]
Ok, well I think I'll leave it for another PR then if any problems arise. The types are working for completion and linting in my use of the editor locally at least.

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [inline bindings/wasm/examples/public/examples.js:267]: No need for `.` in JS - everything is a double - even integers!

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [inline bindings/wasm/examples/public/examples.js:338]: 👍

File: bindings/wasm/examples/public/examples.js
Comment (elalish) [inline bindings/wasm/examples/public/examples.js:27]: I appreciate the idea, but I'm intentionally keeping this first landing-page example super simple. Let's revert this for now. We should add a cool example that showcases more of the CrossSection API though at some point.

File: bindings/wasm/helpers.cpp
Comment (elalish) [inline bindings/wasm/helpers.cpp:160]: Let's do this in a separate PR, but I agree with @pca006132 that instead of `arc_tolerance` we should use our `setMinCircularAngle` etc. functions to define this.

**Review by elalish — CHANGES_REQUESTED** [review]
Just some minor things and then I think we can merge. Let's try to avoid any more functional changes to this PR so we can close the review cycle. We'll need another one before release anyway to address the offset API.

File: bindings/wasm/examples/public/examples.js
Comment (geoffder) [reply bindings/wasm/examples/public/examples.js:267]: Ah ocaml muscle memory acting up again.

File: bindings/wasm/examples/worker.ts
Comment (elalish) [inline bindings/wasm/examples/worker.ts:84]: I must have missed this in code review - is this intentional? Just want to make sure I can safely remove it now.

**Comment by hrgdavor** [comment]
love the addition of slice :)

**Pull request merged**
